### ISSUE TITLE
memberlist can't recover from partitioned groups.

### ISSUE THREAD

**Issue by liuyuan10** [metallb/metallb#589]
Take the most simple two nodes case as an example, if the two are disconnected, they both think the other one is dead and removed them from the list. However, I notice they can not recover from that. Both nodes don't try to connect the other anymore.

I faked the disconnection by:
sudo iptables -A INPUT -p tcp --dport 7946 -j DROP
sudo iptables -A INPUT -p udp --dport 7946 -j DROP
sudo iptables -A OUTPUT -p tcp --dport 7946 -j DROP

From the speaker log, I can see the other node was considered failed but never retried again. They end up in an inconsistent state where both nodes are announcing the same set of IPs.

Related speaker log:
{"caller":"main.go:216","component":"MemberList","msg":"state.go:368: [DEBUG] memberlist: Failed ping: cluster-1-7ccff658b6-l2vrn (timeout reached)","ts":"2020-04-17T21:44:47.010618029Z"}
{"caller":"main.go:216","component":"MemberList","msg":"state.go:459: [INFO] memberlist: Suspect cluster-1-7ccff658b6-l2vrn has failed, no acks received","ts":"2020-04-17T21:44:47.760482541Z"}
{"caller":"main.go:216","component":"MemberList","msg":"state.go:1157: [INFO] memberlist: Marking cluster-1-7ccff658b6-l2vrn as failed, suspect timeout reached (0 peer confirmations)","ts":"2020-04-17T21:44:47.76110408Z"}
{"caller":"main.go:200","msg":"Node event","node addr":"21.0.114.215","node event":"NodeLeave","node name":"cluster-1-7ccff658b6-l2vrn","ts":"2020-04-17T21:44:47.76120179Z"}

After that, no protocol logs show up anymore after above iptables rules are removed.

This is a serious bug. An isolated node will start to announce all services IPs after reconnected causing disruptions to every services in the cluster and it can't recover from that.

Is there way to tell memberlist to keep probing a failed node until it's removed from k8s resources?

**Comment by champtar** [comment]
Thanks a lot for your awesome QA job @liuyuan10 
I have an idea for this one that would also fix some other issue:
Have a goroutine watch the speakers pod (maybe we need to had a headless service), rejoin when a pod is in the api but not in the memberlist, and have the number N of speaker currently up according to the k8s API, so if we have less than N/2 in our memberlist, don't advertise

**Comment by liuyuan10** [comment]
On Fri, Apr 17, 2020 at 5:20 PM Etienne Champetier <notifications@github.com>
wrote:

> Thanks a lot for your awesome QA job @liuyuan10
> <https://github.com/liuyuan10>
> I have an idea for this one that would also fix some other issue:
> Have a goroutine watch the speakers pod (maybe we need to had a headless
> service), rejoin when a pod is in the api but not in the memberlist, and
> have the number N of speaker currently up according to the k8s API, so if
> we have less than N/2 in our memberlist, don't advertise
>
regarding rejoining, you need to consistently check and rejoin. I don't see
there is an event to trigger it which means you need to do that in a loop.

I don't think the 2nd idea works. Say there are 3 nodes and two of those
are dead, you still want to announce from this single instance

**Comment by champtar** [comment]
After some time the 2 dead nodes will be dead in K8S api also, so the last node will update the cluster size to 1 and start announcing again, it'll take a bit of time, but losing more than half your nodes should never happen IMO, so if it takes 40s I think it's fine

**Comment by liuyuan10** [comment]
In the case of 3 node clusters, I don't see loosing two nodes are never
possible. For example, a physical machine dies that takes away all two VMs
on that machine. Also it doesn't help the two node case.

**Comment by rata** [comment]
@liuyuan10 thanks for the report!

Can you confirm if this doesn't happen if you disable memberlist? (just in case, you can disabling as explained [here](https://metallb.universe.tf/release-notes/#version-0-9-2)).

@champtar I wonder if we should disable mlist by default until this issue is fixed (and release 0.9.4 soon). What do you think?

**Comment by champtar** [comment]
@rata I'll try to work on this tomorrow, please give me a bit of time before disable by default

**Comment by rata** [comment]
@champtar no problem! Just wanted to know what you think. We can of course wait and release a bugfix, that is actually better. Thanks! :)

**Comment by johananl** [comment]
Reading the above, this looks to me like a classic [split-brain](https://en.wikipedia.org/wiki/Split-brain_(computing)) scenario. AFAICT there are generally two ways to address this kind of issues:

- Let the two "sides" of the network partition work normally.
- Enforce a quorum-based logic to avoid "competition".

The former may be problematic as it can cause competing ARP replies. The latter typically requires an odd number of memberlist participants to guarantee majority.

I'm still learning this issue (and memberlist...) but I thought it could be useful to frame the discussion using the terms above.

**Comment by johananl** [comment]
I was able to confirm that when isolating two workers, both of them respond to ARP requests simultaneously:

```
sudo tcpdump -i docker0 arp and host 172.17.255.1
```
```
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on docker0, link-type EN10MB (Ethernet), capture size 262144 bytes
19:47:54.271451 ARP, Request who-has 172.17.255.1 tell redacted, length 28
19:47:54.272346 ARP, Reply 172.17.255.1 is-at 02:42:ac:11:00:04 (oui Unknown), length 46
19:47:54.272398 ARP, Reply 172.17.255.1 is-at 02:42:ac:11:00:02 (oui Unknown), length 46
```

I was able to reproduce the duplicate announcement scenario also **with memberlist disabled**. I blocked traffic between a node and the API server by running the following on the node:

```
iptables -I INPUT -p tcp --sport 6443 -j DROP
iptables -I OUTPUT -p tcp --dport 6443 -j DROP
```

After some seconds, the node became unhealthy and the 2nd node started announcing, **however the first node continued to announce, too**.

So, as far as I can tell we have two separate issues:

- Split brain situations.
- memberlist convergence following a network partition.

AFAICT, introducing memberlist into MetalLB was meant to **reduce convergence time**, which it does. memberlist support wasn't meant to address split brain scenarios, nor did it _cause_ them, so I'd say we don't have a regression here either.

I've done some archaeology and found https://github.com/metallb/metallb/issues/195. Looks like there were already discussions around split brain and it was decided to tolerate it as the lesser evil since other options have worse availability overall (e.g. having a "leader" concept makes all traffic drop if the k8s control plane goes down).

Therefore, I suggest we **ignore** the split brain scenario in the context of this issue since it doesn't seem tied to memberlist. If someone has an idea regarding how to address split brain situations without making availability worse (again, see https://github.com/metallb/metallb/issues/195), let's do it in a _new issue_.

Regarding the memberlist re-convergence, IIUC the desired behavior following a network partition is that when the network problem is resolved, the nodes should re-discover each other and the duplicate ARPs should disappear (right?). If that's the case, let's focus the discussion and any proposed solutions on this narrower problem.

Were we able to verify that memberlist indeed doesn't re-converge following a network partition? If so, do we understand exactly why it doesn't? I like the idea of a headless service instead of the current mechanism which relies on querying for all speaker pods by label, however I don't see much difference in terms of re-convergence since a headless service is just another mechanism to get a list of IP addresses. If someone could shed some light on that, that would be great. I will look into how memberlist operates in a bit more detail and maybe then I'll have something more intelligent to say about the re-convergence problem.

**Comment by champtar** [comment]
@johananl everything should be fixed with #595 ;)
1) We need the list of speakers for the initial discovery of other MemberList members, and to rejoin, because MemberList behavior is to not retry after timeout (my initial PR that is now in 0.9.3 was based on some wrong assumption)
2) After a split brain it's not enough to just stop one of the 2 ARP/NDP responders, the switches might have 2 conflicting information, so the new unique leader/responder need to do some gratuitous advertisement. The tricky part is that the new unique leader might think he always was the unique leader.

**Comment by liuyuan10** [comment]
Split brain problem is hard to solve. I can see it's also a problem for VRRP protocol. The major downside I see of it is after network is restored, gateway is caching the wrong MAC so still sends traffic to the wrong node. While it still works, traffic might gets blackholed until next time GW refreshes ARP entry. It might get worse is ARP suppression is enabled in the switches. It's more likely to happen during a rolling network maintenance.

It's nice if we can send GARP even after some time so that at least we know if will be fixed. I agree it's a different issue from the 2nd one (memberlist convergence) which is more important.
